### PR TITLE
Add distortion to voting profiles

### PR DESCRIPTION
**Pull request by esion-liu**

### CODE REVIEW

File: compsoc/evaluate.py [hunk 1]
```diff
@@ -105,6 +106,8 @@ def evaluate_voting_rules(num_candidates: int,
     :rtype: dict[str, dict[str, float]]
     """
     profile = get_profile_from_model(num_candidates, num_voters, voters_model)
+    profile = generate_distorted_from_normal_profile(profile, distort_rate)
+    print(profile.pairs)
```
Comment (asoderlind) [inline compsoc/evaluate.py:110]: remove print statement

File: compsoc/profile.py [hunk 2]
```diff
@@ -185,7 +194,7 @@ def __calc_net_preference(self):
                         preferences.append(pref)  # save preference
                 # Computes the preference
                 preference = sum(preferences)
-                # Save preferences
+                # Save preferencess
```
Comment (asoderlind) [inline compsoc/profile.py:197]: typo

File: compsoc/voter_model.py [hunk 3]
```diff
@@ -122,6 +122,39 @@ def get_pairs_from_model(num_candidates: int, num_voters: int, voters_model: str
...
+    distort a normal profile to generate a distorted profile, distort_rate is from 0. to 1.
+    0 mesns no ditortion at all, and 1 means all ballots only keeps the first candidate.
+    """
+    num_candidates = len(origin_profile.candidates)
+    num_to_remain = round(num_candidates * (1 - distort_rate))
+    if num_to_remain == 0:
+        num_to_remain += 1
+    ballots = origin_profile.pairs.copy()
+    results = []
+    for pattern in ballots:
```
Comment (asoderlind) [inline compsoc/voter_model.py:136]: more clear to use consistent syntax like "for b in ballots" or "for ballot in ballots" I don't think we use "pattern" anywhere else in the SDK

**Review by asoderlind — CHANGES_REQUESTED** [review]

**Comment by asoderlind** [comment]
Also please add some real tests to the `tests/test_profile.py` file, or make a new one called `test_distortion.py` so we can see that it works as expected

**Comment by asoderlind** [comment]
I also think it would be good to add a method `distort(rate == 0.0)` to the Profile class, so that you can generate a profile and then distort that same object easily, would help with the implementation. @YuchenL-KyotoU

**Comment by YuchenL-KyotoU** [comment]
OK I will do it afternoon after class.

iPhoneから送信

2023/06/28 8:17、Axel Söderlind ***@***.***>のメール:

﻿

I also think it would be good to add a method distort(rate == 0.0) to the Profile class, so that you can generate a profile and then distort that same object easily, would help with the implementation. @YuchenL-KyotoU<https://github.com/YuchenL-KyotoU>

**Pull request merged**